controllers/k8ssandra/k8ssandracluster_controller_test.go

```diff
@@ -1195,6 +1198,114 @@ func createMultiDcClusterWithStargate(t *testing.T, ctx context.Context, f *fram
+// Such change is prohibited and should fail
+func changeNumTokensValue(t *testing.T, ctx context.Context, f *framework.Framework, namespace string) {
+	require := require.New(t)
+
+	k8sCtx := "cluster-1"
+
+	kc := &api.K8ssandraCluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      "test",
+		},
+		Spec: api.K8ssandraClusterSpec{
+			Cassandra: &api.CassandraClusterTemplate{
+				Cluster: "test",
+				CassandraConfig: &api.CassandraConfig{
+					CassandraYaml: api.CassandraYaml{
+						NumTokens: pointer.Int(16),
+					},
+				},
+				Datacenters: []api.CassandraDatacenterTemplate{
+					{
+						Meta: api.EmbeddedObjectMeta{
+							Name: "dc1",
+						},
+						K8sContext:    k8sCtx,
+						Size:          1,
+						ServerVersion: "3.11.10",
+						StorageConfig: &cassdcapi.StorageConfig{
+							CassandraDataVolumeClaimSpec: &corev1.PersistentVolumeClaimSpec{
+								StorageClassName: &defaultStorageClass,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	err := f.Client.Create(ctx, kc)
+	require.NoError(err, "failed to create K8ssandraCluster")
+
+	verifyFinalizerAdded(ctx, t, f, client.ObjectKey{Namespace: kc.Namespace, Name: kc.Name})
+
+	verifySuperuserSecretCreated(ctx, t, f, kc)
+
+	verifyReplicatedSecretReconciled(ctx, t, f, kc)
+
+	verifySystemReplicationAnnotationSet(ctx, t, f, kc)
+
+	t.Log("check that the datacenter was created")
+	dcKey := framework.ClusterKey{NamespacedName: types.NamespacedName{Namespace: namespace, Name: "dc1"}, K8sContext: k8sCtx}
+	require.Eventually(f.DatacenterExists(ctx, dcKey), timeout, interval)
+
+	lastTransitionTime := metav1.Now()
+
+	t.Log("update datacenter status to scaling up")
+	err = f.PatchDatacenterStatus(ctx, dcKey, func(dc *cassdcapi.CassandraDatacenter) {
+		dc.SetCondition(cassdcapi.DatacenterCondition{
+			Type:               cassdcapi.DatacenterScalingUp,
+			Status:             corev1.ConditionTrue,
+			LastTransitionTime: lastTransitionTime,
+		})
+	})
+	require.NoError(err, "failed to patch datacenter status")
+
+	t.Log("update datacenter status to ready")
+	err = f.PatchDatacenterStatus(ctx, dcKey, func(dc *cassdcapi.CassandraDatacenter) {
+		lastTransitionTime = metav1.Now()
+		dc.SetCondition(cassdcapi.DatacenterCondition{
+			Type:               cassdcapi.DatacenterReady,
+			Status:             corev1.ConditionTrue,
+			LastTransitionTime: lastTransitionTime,
+		})
+		dc.SetCondition(cassdcapi.DatacenterCondition{
+			Type:               cassdcapi.DatacenterScalingUp,
+			Status:             corev1.ConditionFalse,
+			LastTransitionTime: lastTransitionTime,
+		})
+	})
+	require.NoError(err, "failed to patch datacenter status")
+
+	// Update the datacenter with a different num_tokens value and check that it failed
+	initialNumTokens := kc.Spec.Cassandra.CassandraConfig.CassandraYaml.NumTokens
+	kcKey := framework.ClusterKey{NamespacedName: types.NamespacedName{Namespace: namespace, Name: "test"}, K8sContext: k8sCtx}
```

Comment:
Because the K8ssandraCluster is always (and has to be) created in the control plane cluster, I just use a `client.ObjectKey` which is an alias for `types.NamespacedName`. We have a small helper function, `utils.GetKey(kc)` that I like to use. 

Since the K8ssandraCluster is always in the control plane cluster I typically use `f.Client` for working with it. Maybe it would be better to rename that to `f.ControlPlaneClient`, although that might be a bit verbose 🤔

Reply:
Gotcha, I'll use `cluster-0` only for this test, which makes things simpler.